engine/lib/notification.php

```diff
@@ -322,18 +309,19 @@ function _elgg_notifications_smtp_thread_headers($hook, $type, $returnvalue, $pa
 			$mt = microtime(true);
 			$messageId = "{$urlPath}.entity.{$object->guid}.$mt@{$hostname}";
 		}
-		$returnvalue['headers']["Message-ID"] = $messageId;
-		$container = $object->getContainerEntity();
+
+		$email->addHeader("Message-ID", $messageId);
 
 		// let's just thread comments by default
-		if (($container instanceof \ElggEntity) && ($object instanceof \ElggComment)) {
+		$container = $object->getContainerEntity();
+		if ($container instanceof \ElggEntity && $object instanceof \ElggComment) {
 			$threadMessageId = "<{$urlPath}.entity.{$container->guid}@{$hostname}>";
```

Comment:
I think so, as there is no specific handlers for `In-Reply-To` and `References` headers, they should be included.